docker/image.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as pulumi from "@pulumi/pulumi";
+import { buildAndPushImage, DockerBuild, Registry } from "./docker";
+
+/**
+ * Arguments for constructing an Image resource.
+ */
+export interface ImageArgs {
+    /**
+     * The qualified image name that will be pushed to the remote registry.  Must be a supported image name for the
+     * target registry user.
+     */
+    imageName: pulumi.Input<string>;
+    /**
+     * The Docker build context, as a folder path or a detailed DockerBuild object.
+     */
+    build: pulumi.Input<string | DockerBuild>;
+    /**
+     * The docker image name to build locally before tagging with imageName.
+     */
+    localImageName?: pulumi.Input<string>;
+    /**
+     * Docker registry to push to.
+     */
+    registry: pulumi.Input<ImageRegistry>;
```

Comment:
Is there ever a scenario where you want to build but not push?

Reply:
I can't think of one in any current style of Pulumi workflow - the whole point of this is to make an image available in some external registry that other resources can reference.

If we identify a use case in the future though, adding a `push: false` option here would be easy. 

BTW - Related to `push: false` would be `pull: true` that pulls an image locally instead.  If there's a reason to want to ensure an image exists locally, `pull: true` is going to be a common case for that.  Either way - both of these can be added later once motivating scenarios are more clear.